cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,960 @@
+        for i, (stage_name, stage_) in enumerate(self._stages_dict.items()):
+            if first_stage_num is not None and i < first_stage_num:
+                stage_.skipped = True
+                stage_.required = False
+                logger.info(f'Skipping stage {stage_name}')
+                continue
+
+            if last_stage_num is not None and i > last_stage_num:
+                stage_.skipped = True
+                stage_.required = False
+                continue
+
+            for reqcls in stage_.required_stages_classes:
+                # breakpoint()
+                if reqcls.__name__ in self._stages_dict:
+                    reqstage = self._stages_dict[reqcls.__name__]
+                elif reqcls.__name__ in additional_stages_dict:
+                    reqstage = additional_stages_dict[reqcls.__name__]
+                else:
+                    # stage is not initialized, so we defining it as skipped
+                    reqstage = reqcls(self)
+                    reqstage.skipped = True
+                    additional_stages_dict[reqcls.__name__] = reqstage
+                reqstage.required = True
+                if reqstage.skipped:
+                    logger.info(
+                        f'Stage {reqstage.name} is skipped, '
+                        f'but the output will be required for the stage {stage_.name}'
+                    )
+                stage_.required_stages.append(reqstage)
+        
+        if additional_stages_dict:
+            logger.info(
+                f'Additional required stages added as "skipped": '
+                f'{additional_stages_dict.keys()}'
+            )
+            for name, stage_ in self._stages_dict.items():
+                additional_stages_dict[name] = stage_    
+            self._stages_dict = additional_stages_dict
+
+        logger.info(f'Setting stages: {", ".join(s.name for s in stages if not s.skipped)}')
+        required_skipped_stages = [s for s in stages if s.skipped and s.required]
+        if required_skipped_stages:
+            logger.info(
+                f'Skipped stages that are used to get inputs: '
+                f'{", ".join(s.name for s in required_skipped_stages)}'
+            )
+
+        # Second round - actually adding jobs from the stages.
+        for i, (_, stage_) in enumerate(self._stages_dict.items()):
+            if not stage_.skipped:
+                logger.info(f'*' * 60)
+                logger.info(f'Stage {stage_.name}')
+
+            if stage_.required:
+                logger.info(f'Adding jobs for stage {stage_.name}')
+                stage_.output_by_target = stage_.add_to_the_pipeline(self)
+
+            if not stage_.skipped:
+                logger.info(f'')
+                if last_stage_num and i >= last_stage_num:
+                    logger.info(f'Last stage is {stage_.name}, stopping here')
+                    break
+
+    def get_datasets(self, only_active: bool = True) -> list[Dataset]:
```

Comment:
This feeling like it pollutes that Pipeline definition, but I don't have a good grasp.